### PR TITLE
fetch

### PR DESCRIPTION
**Pull request by IdoGvili**
Switched from local data.json file to fetching fromhttps://raw.githubusercontent.com/IdoGvili/TaskList-repo/master/src/Data.json

### CODE REVIEW

File: src/App.jsx [hunk 1]
```diff
@@ -1,12 +1,25 @@
...
 
 import Header from './Header';
 import ToDoList from './ToDoList';
 import ToDoForm from './ToDoForm';
 
 function App() {
-    const [toDoList, setToDoList] = useState(Data);
+    const [toDoList, setToDoList] = useState([]);
+
+    useEffect(() => {
```
Comment (SaharAvr) [inline src/App.jsx]: awesome. just move the `fetchData` function outside of  the `useEffect` hook.
and bonus task - try to show a loader in the UI while the data is fetching (while developing the feature you can mock the loading time by awaiting an empty promise with setTimeout)

**Comment by SaharAvr** [comment]
also i like that your commit messages are short, but they are a bit too short. - from now on try to make your commit messages a bit more descriptive (for example for this one the message could have been "fetch todo-list data remotely")

File: src/App.jsx [hunk 2]
```diff
@@ -1,23 +1,27 @@
 import React, { useState, useEffect } from 'react';
+import { ColorRing } from 'react-loader-spinner';
 import data from './data.json';
 
 import Header from './Header';
 import ToDoList from './ToDoList';
 import ToDoForm from './ToDoForm';
 
+async function fetchData(setToDoList) {
```
Comment (SaharAvr) [inline src/App.jsx]: i meant move it outside of useEffect, but keep it inside the component(:
let's move this inside the component. you may use `useCallback` hook (read about it and let me know if you need more explanation about it)

in general it is not recommended to overuse useCallbacks (some suggest not to use it at all), but we are doing it for the sake of getting to know React better

File: src/App.jsx [hunk 3]
```diff
@@ -1,23 +1,27 @@
...
+    await new Promise((resolve) => {
```
Comment (SaharAvr) [inline src/App.jsx]: you forgot to remove this

File: src/App.jsx (hunk 2 above)
Comment (SaharAvr) [inline src/App.jsx]: also, since you are fetching the todo list, let's move this fetching function into the TodoList component

File: src/App.jsx [hunk 4]
```diff
@@ -59,10 +63,32 @@ function App() {
     };
 
     return (
-        <div className="App">
-            <Header />
-            <ToDoList toDoList={toDoList} handleButton={handleButton} />
-            <ToDoForm addTask={addTask} />
+        <div>
+            {toDoList.length === 0 ? (
```
Comment (SaharAvr) [inline src/App.jsx]: same, this as well should be implemented inside TodoList component

File: src/App.jsx [hunk 5]
```diff
@@ -22,7 +30,7 @@ function App() {
         setToDoList(filtered);
     };
 
-    const addTask = (userInput) => {
+    const onAddTodo = (userInput) => {
```
Comment (SaharAvr) [inline src/App.jsx]: instead of `userInput` let's change the parameters to `({ task, date })`

File: src/App.jsx [hunk 6]
```diff
@@ -1,17 +1,25 @@
...
+                <Todos />
+            </div>
+        </div>
+    );
+}
+function Todos() {
+    const [toDoList, setToDoList] = useState([]);
 
-    const handleButton = (id) => {
+    const onRemoveTodo = (id) => {
```
Comment (SaharAvr) [inline src/App.jsx]: to be persistent with the `onAddTodo` function parameters (which now accepts task), let's also expect a tasks here (then you can grab its id). so `({ task })` or (task) as parameters

File: src/App.jsx [hunk 7]
```diff
@@ -33,24 +41,38 @@ function App() {
...
     return (
-        <div className="App">
-            <Header />
-            <ToDoList toDoList={toDoList} handleButton={handleButton} />
-            <ToDoForm addTask={addTask} />
-        </div>
+        <>
+            <ToDoList
+                toDoList={toDoList}
+                setToDoList={setToDoList}
```
Comment (SaharAvr) [inline src/App.jsx]: why do you need `setToDoList` here? remove it

File: src/App.jsx [hunk 8]
```diff
@@ -33,24 +41,38 @@ function App() {
...
 
     return (
-        <div className="App">
-            <Header />
-            <ToDoList toDoList={toDoList} handleButton={handleButton} />
-            <ToDoForm addTask={addTask} />
-        </div>
+        <>
+            <ToDoList
+                toDoList={toDoList}
```
Comment (SaharAvr) [inline src/App.jsx]: this is ok. i would rename to `todos`

File: src/ToDoList.jsx [hunk 9]
```diff
@@ -1,14 +1,32 @@
 import React from 'react';
+import { ColorRing } from 'react-loader-spinner';
 import ToDo from './ToDo';
 
-function ToDoList({ toDoList, handleButton }) {
-    return (
+export default function ToDoList({ toDoList, onRemoveTodo }) {
+    return toDoList.length === 0 ? (
```
Comment (SaharAvr) [inline src/ToDoList.jsx]: well, what if i had todos, then deleted myself all of them. the loader would still appear...
you need a different way to indicate whether or not to show a loader. you could use a loader prop, or set a different initial state for the list (for example you could set it to "null" and present the loader only if its value is null)

File: src/App.jsx [hunk 10]
```diff
@@ -1,17 +1,25 @@
...
+                <Todos />
+            </div>
+        </div>
+    );
+}
+function Todos() {
+    const [toDoList, setToDoList] = useState(null);
 
-    const handleButton = (id) => {
+    const onRemoveTodo = ({ id }) => {
```
Comment (SaharAvr) [inline src/App.jsx]: i suppose `{ id }` is the deconstruction of `task` paramater but it is not very clear. How about naming it taskId or not deconstructing at all?

Also, let's choose one - "task" or "todo" and align all related namings accordingly

File: src/App.jsx [hunk 11]
```diff
@@ -56,7 +56,7 @@ function Todos() {
         const fetchedData = await response.json();
         const newData = fetchedData.map((task) => ({ ...task, id: v4() }));
         setToDoList(newData);
-    }, [setToDoList]);
+    }, []);
```
Comment (IdoGvili) [inline src/App.jsx:59]: removed setToDoList from dependencies so the function won't change

**Review by SaharAvr — APPROVED** [review]

**Comment by IdoGvili** [comment]
Now fetches data from website instead of a file, uuid instead sequential id, refactored objects in toDoList to always be called toDo, 'add and remove from list' functions are now in Todos component instead of App

**Pull request merged**